Add support for iterators on STRING types

big-andy-coates opened the pull request:
KSQL requires this. It was added to the KSQL specific version in the following PR:

https://github.com/dguy/ksql/commit/e8e930ace6eac65db6cf1fcc79352600d69767ce

But switching KSQL to use this library has caused this to fail.

@C0urante would be great to get this merged and a new release released :D

C0urante commented:
@big-andy-coates, @dguy sorry for missing this! I noticed the bug fix that that PR brought for correcting iterative behavior but somehow the addition of string iteration made it by me.

I chose to tackle the iterator behavior by using an `IdentityHashMap<Schema, Iterator<Object>>` instead of including Schema field names, as it seemed more readable and added less complexity to the code base. Would it be possible to alter this PR so that only the string iterator behavior is added, without using field names to differentiate between schemas? I'd be happy to try to make the changes myself if that'd be easier.

C0urante commented:
@big-andy-coates After examining the PR that was merged for KSQL, it looks like there's a difference in behavior for String iteration between that one and the one proposed here. In the [merged KSQL PR](https://github.com/dguy/ksql/commit/e8e930ace6eac65db6cf1fcc79352600d69767ce#diff-39600f78a6093d4a798babbb44001661R885), String iteration just iterates over an integer and converts it to a String; however, in [this PR](https://github.com/confluentinc/avro-random-generator/pull/2/files#diff-999b8ded4e0d870d4d2f8097b339b47bR898) it looks like the name of the field for the schema is appended along with an underscore to the beginning of that String.

I've created a separate branch that adds support for String iteration and adds additional `prefix` and `suffix` fields for String schemas. I believe the functionality there will be enough to suit your needs as long as you don't mind specifying a `prefix` for strings you'd like to generate with a format like `name_1`, `name_2`, etc. Let me know how that sounds!

big-andy-coates commented:
Hi @C0urante,

Thanks for your feedback!

I've updated from your branch to add prefix and suffix support, and reverted the use of fieldname in the iteration cache.

I've switched the code to create a schema parser on the fly, as they hold internal state and throw errors if you try and parse the same schema twice. (causing tests to fail). 

Should be good to go now.

dguy commented on src/test/java/io/confluent/avro/random/generator/GeneratorTest.java:
```diff
@@ -20,29 +22,31 @@
     final Generator generator = new Generator(content, RNG);
-    generator.generate();
+    final Object generated = generator.generate();
+    System.out.println(fileName + ": " + generated);
```
did you mean to leave this?

big-andy-coates replied:
I did yes, mainly because the tests at the moment don't actually test the output is correct - just that they don't blow up.  So this printlin at least allows a dev to see the output. (Plus I added explicit tests around the bits I added).
